clouddriver-sql/src/main/kotlin/com/netflix/spinnaker/clouddriver/sql/event/SqlEventRepository.kt

```diff
@@ -0,0 +1,298 @@
+import com.netflix.spinnaker.clouddriver.event.Aggregate
+import com.netflix.spinnaker.clouddriver.event.CompositeSpinnakerEvent
+import com.netflix.spinnaker.clouddriver.event.EventMetadata
+import com.netflix.spinnaker.clouddriver.event.SpinnakerEvent
+import com.netflix.spinnaker.clouddriver.event.exceptions.AggregateChangeRejectedException
+import com.netflix.spinnaker.clouddriver.event.persistence.EventRepository
+import com.netflix.spinnaker.clouddriver.event.persistence.EventRepository.ListAggregatesCriteria
+import com.netflix.spinnaker.clouddriver.sql.transactional
+import com.netflix.spinnaker.config.ConnectionPools
+import com.netflix.spinnaker.kork.sql.config.SqlProperties
+import com.netflix.spinnaker.kork.sql.config.SqlRetryProperties
+import com.netflix.spinnaker.kork.sql.routing.withPool
+import com.netflix.spinnaker.kork.version.ServiceVersion
+import de.huxhorn.sulky.ulid.ULID
+import io.github.resilience4j.retry.Retry
+import io.github.resilience4j.retry.RetryConfig
+import io.github.resilience4j.retry.RetryRegistry
+import org.jooq.Condition
+import org.jooq.DSLContext
+import org.jooq.impl.DSL.currentTimestamp
+import org.jooq.impl.DSL.field
+import org.jooq.impl.DSL.table
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.validation.Validator
+import java.time.Duration
+import java.util.Optional
+import java.util.UUID
+
+class SqlEventRepository(
+  private val jooq: DSLContext,
+  sqlProperties: SqlProperties,
+  private val serviceVersion: ServiceVersion,
+  private val objectMapper: ObjectMapper,
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val registry: Registry,
+  private val validator: Validator,
+  private val retryRegistry: RetryRegistry
+) : EventRepository {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+  private val retryProperties: SqlRetryProperties = sqlProperties.retries
+
+  private val eventCountId = registry.createId("eventing.events")
+
+  /**
+   * TODO(rz): Performance.
+   */
+  override fun save(
+    aggregateType: String,
+    aggregateId: String,
+    originatingVersion: Long,
+    newEvents: List<SpinnakerEvent>
+  ) {
+    val eventNames = newEvents.joinToString { it.javaClass.simpleName }
+    log.debug("Saving $aggregateType/$aggregateId expecting version $originatingVersion with [$eventNames]")
+
+    val aggregateCondition = field("aggregate_type").eq(aggregateType)
+      .and(field("aggregate_id").eq(aggregateId))
+
+    // TODO(rz): Get this from Spring?
+    val retry = RetryConfig.custom<Retry>()
+      .maxAttempts(retryProperties.transactions.maxRetries)
+      .waitDuration(Duration.ofMillis(retryProperties.transactions.backoffMs))
+      .ignoreExceptions(AggregateChangeRejectedException::class.java)
+      .build()
+
+    try {
+      withPool(POOL_NAME) {
+        jooq.transactional(retryRegistry.retry("eventSave", retry)) { ctx ->
```

Comment:
Potentially long running transaction here (esp. with potential for retries in `getNextMonotonicToken`), be sure to profile under heavy concurrent load.